### ISSUE TITLE
Token retractions may not occur when an accumulator reduces to nil

### ISSUE THREAD

**Issue by mrrodriguez** [rbrush/clara-rules#78]
I have opted to bring this up as an issue instead of submit a pull request for this one because I wanted to get some feedback first.

I have a preliminary patch exposing this issue and a fix (that I'd likely clean up if we chose to go with it) @ https://github.com/mrrodriguez/clara-rules/compare/accum-reduced-nil-retract?expand=1.

## The Problem ##

I noticed some strange behavior in accumulating rules in Clara when an accumulator reduced to a nil value at one point, but later had a new fact inserted which should have retracted the previously reduced value from the working memory.

There are several cases where this can show up, but here I'll just paste the test cases I came up in the link/patch I have above.

Let's use an accumulator that may return a nil value when performing the `:reduce-fn`:
```clj
(def maybe-nil-max-temp (accumulate :reduce-fn (fn [value item]
                                                 (let [t (:temperature item)]
                                                   (if (and t ; No :temperature just returns `value`
                                                            (or (= value nil)
                                                                (> t (:temperature value))))
                                                     item
                                                     value)))))
```

The first test case is showing that a given session can behave differently depending on the order that facts were inserted.  If at first a fact was inserted that caused a nil accumulator reduced value, the later fact that is inserted does not properly "clean up" the tokens propagated by the previous accumulator reduction.

Flipping the order works since the accumulator `maybe-nil-max-temp` will not return nil anymore.
```clj
(deftest test-nil-accum-reduced-has-activation-retracted-when-new-item-inserted
  (let [coldest-temp-rule (dsl/parse-rule [[?coldest <- maybe-nil-max-temp :from [Temperature]]]
                                          (insert! (->Cold ?coldest)))
        coldest-temp-query (dsl/parse-query [] [[?cold <- Cold]])
        temp-nil (->Temperature nil "MCI")
        temp-10 (->Temperature 10 "MCI")
        
        session (mk-session [coldest-temp-rule coldest-temp-query])
        insert-nil-first-session (-> session
                                     (insert temp-nil)
                                     (insert temp-10)
                                     fire-rules)
        insert-nil-second-session (-> session
                                      (insert temp-10)
                                      (insert temp-nil)
                                      fire-rules)]

    (is (= (count (query insert-nil-first-session coldest-temp-query))
           (count (query insert-nil-second-session coldest-temp-query))))

    (is (= #{{:?cold (->Cold temp-10)}}
           (set (query insert-nil-first-session coldest-temp-query))
           (set (query insert-nil-second-session coldest-temp-query))))))
```
In general, I would not expect fact insertion order to result in different outcomes with the session setup above.

--
The second case is simpler.  Insert a fact that causes an accumulator to reduce to a nil value, then retract that fact.  The tokens propagated by the initially reduced nil value are not "cleaned up" in working memory (note: using the same accumulator `maybe-nil-max-temp` as before).
```clj
(deftest test-nil-accum-reduced-has-activation-retracted-when-item-retracted
  (let [coldest-temp-rule (dsl/parse-rule [[?coldest <- maybe-nil-max-temp :from [Temperature]]]
                                          (insert! (->Cold ?coldest)))
        coldest-temp-query (dsl/parse-query [] [[?cold <- Cold]])

        nil-temp (->Temperature nil "MCI")
        session (-> (mk-session [coldest-temp-rule coldest-temp-query])
                    (insert nil-temp)
                    (retract nil-temp)
                    fire-rules)]
    
    (is (empty? (set (query session coldest-temp-query))))))
```

## Proposal ##

The issue described above stems from the ambiguity in `clara.rules.memory/get-accum-reduced` returning nil.  The nil can either mean (1) that *nothing was ever* reduced for this node
 (actually `[memory node join-bindings fact-bindings]`), or it means (2) that *there was* a previously reduced result of nil.

To fix this, it seems it would be best to return something other than nil for one of these two cases.  I have chosen to return something different for case (1) - a "marker" keyword.  There are more options available here and it'd likely be nice to have some reusable function for the 
`(not= :clara.rules.memory/no-accum-reduced <item>)` sort of checks.

## Still TODO ##

Once I get some feedback on this, I'd plan to add some more tests to demonstrate this issue+fix for the clara.rules.engine.AccumulateWithJoinFilterNode along with the c.r.e.AccumulateNode tests I already have, since it should also have the same sort of fixes (maybe some of this duplication should be cleaned up eventually).

Also, there are similar functions in `clara.rules.memory/get-accum-reduced-all` and 
`c.r.m/get-accum-reduced-complete`.  I understand only a fraction of their purpose, so I have not dug into them to see if they are affected in a similar way as `c.r.m/get-accum-reduced`.

**Comment by rbrush** [comment]
Nice find! A namespace-qualified keyword to distinguish this condition definitely sounds like the right way to go. 

I think get-accum-reduced-all is indeed subject to the same issue...the only difference between it and get-accum-reduced is get-accum-reduced-all finds all reduced items that match the values that are used in a *join*, where get get-accum-reduced finds items that match both a *join* bindings and the bindings specific to that reduce, but not used in a join.  (The -complete version isn't very well named, but is only used for serializing the state of a session.)

**Comment by mrrodriguez** [comment]
Thanks for the feedback.  I created a pull request with a slightly cleaned up version of the above @ https://github.com/rbrush/clara-rules/pull/79.

--

I looked at `get-accum-reduced-all` to see if I could recreate the issue there.  I am not able to recreate a scenario where it can return a nil that would cause tokens to not be retracted.  I'll note that it is a relatively complex flow to follow so there is a chance that I missed a detail along the way.

--

My reasoning goes likes this:

`get-accum-reduced-all` is used in the following places:
(1) AccumulateNode impls for `clara.rules.engine/left-activate` and `c.r.e/left-retract`
(2) AccumulateWithJoinNode impls for the same `left-activate` and `left-retract`

--

The easier situation is for AccumulateNode (1).
The `c.r.m/get-accum-reduced-all` will return a non-nil if there are any previously reduced values since it returns a map from `fact-bindings` to the accum-reduced value - which may be nil.  Since the map returned is non-nil, the `clara.rules.engine/send-accumulated` and the `c.r.e/retract-accumulated` functions both will be called when expected.  They both seem to correctly handle a nil `result` value passed in.  It looks like the correct token is propagated through the network (for insertions and retractions).

So when a nil is returned in this flow, it really means that there are no previous reductions.

--

In the case of the AccumulateWithJoinNode (2), there is a comment explicitly in the AccumulateWithJoinNode impl of `left-activate`:
```clj
;; This accumulator keeps candidate facts rather than fully reduced values in the working memory,
;; since the reduce operation must occur per token. Since there are no candidate facts
;; in this flow, just put an empty vector into our memory.

```
So the possibly nil value from an accumulated reduction step is not held in working memory for this node.
When a `left-activate` happens here and there are `grouped-candidate-facts` found in working memory that have not yet been passed to the accumulator's `reduce-fn`.  These facts are ran through the `c.r.e/do-accumulate` and the result of that is passed to the `send-accumulated` function that I mentioned above.  This again behaves correctly for nil accum-reduced values.

The `left-retract` impl for AccumulateWithJoinNode will call `do-accumulate`, the same as `left-activate`, on the deferred `candidate-facts` that were held in working memory for this node.  The result of the `do-accumulate` may be a nil value, but this will be passed to the `retract-accumulated` mentioned above.  And again, this seems to work correctly.
